### PR TITLE
feat: :sparkles: added ability to output exitcode in Terraform Plan task

### PR DESCRIPTION
**Pull request by JaydenMaalouf**
This feature allows the exitcode of Terraform Plan to be consumed within the pipeline after a run

### CODE REVIEW

**Comment by mericstam** [comment]
Hi @JaydenMaalouf , for now you need to add the same changes to task.loc.json as you did for task.json.

**Comment by JaydenMaalouf** [comment]
> Hi @JaydenMaalouf , for now you need to add the same changes to task.loc.json as you did for task.json.

Thanks, this has been updated!

**Comment by mericstam** [comment]
Hi, finally got the extension installed in our dev env.
Unfortunately the suggested changes does not work.  Azure Pipelines picks up on terraform exit code and fails. 

image

same for linux and windows agents.

**Comment by JaydenMaalouf** [comment]
> Hi, finally got the extension installed in our dev env. Unfortunately the suggested changes does not work. Azure Pipelines picks up on terraform exit code and fails.
> 
> image
> 
> same for linux and windows agents.

Could we potentially change the exec to ignore the result?
Something like this:
```ts
let result = terraformTool.exec(<IExecOptions> {
    cwd: planCommand.workingDirectory,
    ignoreReturnCode: true
});

if (result === 1) {
    throw new Error('Terraform failed with exit code: ' + result);
}

return result;
```

**Comment by mericstam** [comment]
Yeah that looks good!  was looking through the azure-pipelines-task-lib and found the same. update the code and I will run another round of tests in our dev environment.

**Comment by JaydenMaalouf** [comment]
> Yeah that looks good! was looking through the azure-pipelines-task-lib and found the same. Please update the code and I will run another round of tests in our dev environment.

Awesome, thanks for that!
I've updated the code - decided to add the error message in the loc files for consistency sake too
Hope I haven't missed anything haha

File: Tasks/TerraformTask/TerraformTaskV1/src/base-terraform-command-handler.ts [hunk 1]
```diff
@@ -119,16 +119,23 @@ export abstract class BaseTerraformCommandHandler {
...
+            `${tasks.getInput("commandOptions")} -detailed-exitcode`
         );
         
         let terraformTool;
         terraformTool = this.terraformToolHandler.createToolRunner(planCommand);
         this.handleProvider(planCommand);
     
-        return terraformTool.exec(<IExecOptions> {
-            cwd: planCommand.workingDirectory
+        let result = terraformTool.exec(<IExecOptions> {
```
Comment (mericstam) [inline Tasks/TerraformTask/TerraformTaskV1/src/base-terraform-command-handler.ts:129]: need to await result  
```suggestion
        let result = await terraformTool.exec(<IExecOptions> {
```

File: Tasks/TerraformTask/TerraformTaskV2/src/base-terraform-command-handler.ts [hunk 2]
```diff
@@ -120,16 +120,23 @@ export abstract class BaseTerraformCommandHandler {
...
+            `${tasks.getInput("commandOptions")} -detailed-exitcode`
         );
         
         let terraformTool;
         terraformTool = this.terraformToolHandler.createToolRunner(planCommand);
         this.handleProvider(planCommand);
     
-        return terraformTool.exec(<IExecOptions> {
-            cwd: planCommand.workingDirectory
+        let result = terraformTool.exec(<IExecOptions> {
```
Comment (mericstam) [inline Tasks/TerraformTask/TerraformTaskV2/src/base-terraform-command-handler.ts:130]: Here as well :)
```suggestion
        let result = await  terraformTool.exec(<IExecOptions> {
```

Comment (JaydenMaalouf) [reply Tasks/TerraformTask/TerraformTaskV2/src/base-terraform-command-handler.ts]: Update PR based on comments

File: Tasks/TerraformTask/TerraformTaskV1/src/base-terraform-command-handler.ts (hunk 1 above)
Comment (JaydenMaalouf) [reply Tasks/TerraformTask/TerraformTaskV1/src/base-terraform-command-handler.ts]: Updated PR based on comments

**Comment by mericstam** [comment]
Hi, I am curious if you see the section 'This branch is out-of-date with the base branch' in checks? if you do please click 'update branch' to integrate the latest changes from main before I merge.  If you can't see it, I can do it but since it is your fork it might be better if you do it.

**Comment by JaydenMaalouf** [comment]
> Hi, I am curious if you see the section 'This branch is out-of-date with the base branch' in checks? if you do please click 'update branch' to integrate the latest changes from main before I merge.  If you can't see it, I can do it but since it is your fork it might be better if you do it.   

Yep!
Merged in main into my branch.
Appreciate your time!

**Review by mericstam — APPROVED** [review]
Looks good!

**Comment by mericstam** [comment]
this feature is now released to public (0.1.14)

**Comment by JaydenMaalouf** [comment]
> this feature is now released to public (0.1.14)

Thanks!
I can see the variable is now being set appropriately.
image

How do I reference this in another job? As it seems isOutput=false, which I think is required to use it as a dependency
The best I've got so far is to run a script with the following
```bash
echo "##vso[task.setvariable variable=outputChangesPresent;isOutput=true]$(echo $(changesPresent))```
To reference the variable outside of the scope of the current job.

**Comment by JaydenMaalouf** [comment]
> > this feature is now released to public (0.1.14)
> 
> Thanks! I can see the variable is now being set appropriately. image
> 
> How do I reference this in another job? As it seems isOutput=false, which I think is required to use it as a dependency The best I've got so far is to run a script with the following
> 
> ```shell
> echo "##vso[task.setvariable variable=outputChangesPresent;isOutput=true]$(echo $(terraformPlan.changesPresent))
> ```
> 
> I had to add a name to the task to reference it
> 
> ```
> - task: TerraformTaskV2@2
>   name: terraformPlan
>   displayName: "Terraform: Plan"
> ```
> 
> This allows me to reference the variable in another job - but should this not be possible with the changesPresent variable directly off the task?

@mericstam I dig some digging and we can do this directly in the `task.setVariable`
PR has been raised here: https://github.com/microsoft/azure-pipelines-terraform/pull/41

**Comment by vanmash** [comment]
I am not sure it this is really the same as https://github.com/microsoft/azure-pipelines-extensions/issues/962

there the idea, to me, was to allow to gracefully skip `apply` task in case `plan` didn't detected any changes.

Using exit codes has this disadvantage that the pipeline is not "green" and for example Bitbucket can see it as failed.

Any thoughts? Should I open another issue for it?

**Comment by mericstam** [comment]
Hi @vanmash, [detailed-exitcode](https://www.terraform.io/cli/commands/plan#detailed-exitcode) exits almost as before but adds the exit code 2 which indicates changes present.  In the terraformtask we ignore the returncodes see: [line 133](https://github.com/microsoft/azure-pipelines-terraform/blob/main/Tasks/TerraformTask/TerraformTaskV2/src/base-terraform-command-handler.ts#L133)  and then trap the exit-codes and if it is 0 or 2 we don't cast an error.   see: [line136](https://github.com/microsoft/azure-pipelines-terraform/blob/main/Tasks/TerraformTask/TerraformTaskV2/src/base-terraform-command-handler.ts#L136)   then  on [line 217](https://github.com/microsoft/azure-pipelines-terraform/blob/main/Tasks/TerraformTask/TerraformTaskV2/src/base-terraform-command-handler.ts#L217) we set an outputvariable named changesPresent which you can use in an condition on the apply step.  In my tests we got the build green.  Did you encounter any issues? If so please provide some more information so we can fault trace

**Comment by vanmash** [comment]
Hi @mericstam , thank you. I am like 50% there :)

I added `name:` to the `plan` task. And `condition:` to the `apply` stage.

```yaml
stages:
  - stage: terraform_plan
    jobs:
      - job: plan
        ...
        steps:
        ...
        - task: TerraformTaskV2@2
          name: 'plan'     
          inputs:
            provider: 'azurerm'
            command: 'plan'
        ...

  - stage: terraform_apply
    dependsOn: [terraform_plan]
    condition: eq(dependencies.terraform_plan.outputs['terraform_plan.plan.changesPresent'], 'true')
```

The `apply` stage is skipped when there are no changes. Good.
But it's also skipped for runs *with* changes. Not so good.
Do you see any obvious mistakes?

**Pull request merged**